### DISCUSSION TITLE
[Q] Is RoadRunner queueing incoming requests if all workers are busy at the moment?

### DISCUSSION THREAD

**Discussion by sprymiker** [roadrunner-server/roadrunner#622] [Q&A] [answered]
As example:

I have one worker that processes a request within 1s.
What would it be if I send 2 requests at the same time?

Option 1. 
One request is processed. Second request if failed

Option 2.
Both requests are processed one by one.

Which behavior would it be? And how to configure the behavior?

**Comment by wolfy-j** [comment, marked as answer]
Option 2 is correct. You can control the size of the queue using the `allocate_timeout` option. Default is 1 minute.

**Comment by rustatian** [comment]
> As example:
> 
> I have one worker that processes a request within 1s.
> What would it be if I send 2 requests at the same time?
> 
> Option 1.
> One request is processed. Second request if failed
> 
> Option 2.
> Both requests are processed one by one.
> 
> Which behavior would it be? And how to configure the behavior?

Yeah, as @wolfy-j said, that would be option 2. 
The process will read on the `net.poll` file descriptor. After that, the first request will be sent to Go's http.handler, while the second request (which was sent at the same time) will wait on the `worker.get` operation (net.poll -> handler -> worker.get). Since RR2 container (in your sample) has only 1 worker, this would be blocking operation. When the worker is released and returned back to the container, the second request can safely get it to process its payload.

**Comment by rustatian** [comment]
Simple diagram would be the following:
1 -> net.poll -> http.handler -> workers.pool -> get -> return response   -------------> put the worker
2 -> net.poll -> http.handler -> workers.pool -> get -> oops, no free workers, wait -> get the worker ^^ -> return response -> put the worker

Of course, when you sent requests at the literally same time, the 2-nd request can first get the worker, but this is something, that has indeterministic behavior.

**Reply by wolfy-j** [reply]
This is why RR can handle 1000 concurrency even with 10 workers.

**Comment by sprymiker** [comment]
Thanks, guys!
